src/core/algorithms/od/fastod/fastod.h

```diff
@@ -0,0 +1,189 @@
+    using AttributePair = fastod::AttributePair;
+    using AttributeSet = fastod::AttributeSet;
+    using PartitionCache = fastod::PartitionCache;
+    using DataFrame = fastod::DataFrame;
+    using Timer = fastod::Timer;
+
+    config::TimeLimitSecondsType time_limit_seconds_ = 0u;
+    bool is_complete_ = true;
+    size_t level_ = 1;
+    size_t od_count_ = 0;
+    size_t fd_count_ = 0;
+    size_t ocd_count_ = 0;
+
+    std::vector<AscCanonicalOD> result_asc_;
+    std::vector<DescCanonicalOD> result_desc_;
+    std::vector<SimpleCanonicalOD> result_simple_;
+    std::vector<std::unordered_set<AttributeSet>> context_in_each_level_;
+    std::unordered_map<AttributeSet, AttributeSet> cc_;
+    std::unordered_map<AttributeSet, std::unordered_set<AttributePair>> cs_asc_;
+    std::unordered_map<AttributeSet, std::unordered_set<AttributePair>> cs_desc_;
+
+    PartitionCache partition_cache_;
+
+    AttributeSet schema_;
+    std::shared_ptr<DataFrame> data_;
+    config::InputTable input_table_;
+
+    Timer timer_;
+
+    void MakeExecuteOptsAvailable() override;
+    void LoadDataInternal() override;
+    void ResetState() override;
+    unsigned long long ExecuteInternal() final;
+
+    void PrepareOptions();
+    void RegisterOptions();
+    void MakeLoadOptionsAvailable();
+
+    template <bool ascending>
+    void AddToResult(fastod::CanonicalOD<ascending>&& od) {
+        if constexpr (ascending)
+            result_asc_.emplace_back(std::move(od));
+        else
+            result_desc_.emplace_back(std::move(od));
+    }
+
+    void AddToResult(SimpleCanonicalOD&& od) {
+        result_simple_.emplace_back(std::move(od));
+    }
+
+    bool IsTimeUp() const;
+
+    void CCPut(AttributeSet const& key, AttributeSet attribute_set);
+    AttributeSet const& CCGet(AttributeSet const& key);
+
+    template <bool ascending>
+    void CSPut(AttributeSet const& key, AttributePair const& value) {
+        if constexpr (ascending)
+            cs_asc_[key].emplace(value);
+        else
+            cs_desc_[key].emplace(value);
+    }
+
+    template <bool ascending>
+    void CSPut(AttributeSet const& key, AttributePair&& value) {
+        if constexpr (ascending)
+            cs_asc_[key].emplace(std::move(value));
+        else
+            cs_desc_[key].emplace(std::move(value));
+    }
+
+    template <bool ascending>
+    std::unordered_set<AttributePair>& CSGet(AttributeSet const& key) {
+        if constexpr (ascending)
+            return cs_asc_[key];
+        else
+            return cs_desc_[key];
+    }
+
+    void Initialize();
+    void ComputeODs();
+    void PruneLevels();
+    void CalculateNextLevel();
+
+    template <bool ascending>
+    void AddCandidates(AttributeSet const& context, std::vector<AttributeSet> const& delAttrs) {
+        if (level_ == 2) {
+            for (model::ColumnIndex i = 0; i < data_->GetColumnCount(); i++) {
+                for (model::ColumnIndex j = 0; j < data_->GetColumnCount(); j++) {
```

Comment:
On the page 6(726), right column, is written that we can store only {A,B} instead both with {B,A}. Can we do `for (model::ColumnIndex j = i + 1` here and remove `if (i == j) continue;`?